src/sources/ukt.rs

```diff
@@ -0,0 +1,335 @@
+//! UKT source for us to sync with Famedly Zitadel.
+
+use std::collections::HashMap;
+
+use anyhow::{bail, Context, Result};
+use chrono::Utc;
+use reqwest::Client;
+use serde::Deserialize;
+use url::Url;
+
+use crate::{config::FeatureFlags, Config, FeatureFlag};
+
+/// UKT Source
+pub struct SourceUkt {
+	/// UKT Source configuration
+	source_ukt_config: SourceUktConfig,
+	/// Optional set of features
+	feature_flags: FeatureFlags,
+	/// Reqwest client
+	client: Client,
+}
+
+impl SourceUkt {
+	/// Create a new SourceUkt instance
+	pub fn new(config: &Config) -> Result<Self> {
+		let client = Client::new();
+		Ok(Self {
+			source_ukt_config: match config.source_ukt.clone() {
+				Some(source_ukt) => source_ukt,
+				None => bail!("Endpoint configuration is missing"),
+			},
+			feature_flags: config.feature_flags.clone(),
+			client,
+		})
+	}
+
+	/// Get list of user emails that have been removed
+	pub async fn get_removed_user_emails(&self) -> Result<Vec<String>> {
+		if self.feature_flags.is_enabled(FeatureFlag::DryRun) {
+			tracing::warn!("Not fetching during a dry run");
+			return Ok(vec![]);
+		}
+
+		let oauth2_token = self.get_oauth2_token().await?;
+		let email_list = self.fetch_list(oauth2_token).await?;
+
+		Ok(email_list)
+	}
+
+	/// Get the OAuth2 token
+	async fn get_oauth2_token(&self) -> Result<OAuth2Token> {
+		let mut params = HashMap::new();
+		params.insert("grant_type", &self.source_ukt_config.grant_type);
+		params.insert("scope", &self.source_ukt_config.scope);
+		params.insert("client_id", &self.source_ukt_config.client_id);
+		params.insert("client_secret", &self.source_ukt_config.client_secret);
+
+		let response = self
+			.client
+			.post(self.source_ukt_config.oauth2_url.clone())
+			.form(&params)
+			.send()
+			.await?;
+
+		if response.status() != reqwest::StatusCode::OK {
+			anyhow::bail!("Error in response: {}", response.status())
+		}
+
+		let response: serde_json::Value = response.json().await?;
+
+		if let Some(error) = response.get("error") {
+			anyhow::bail!("Error in response: {}", error)
```

Comment:
To confirm, you prefer me to do this:
```
if let Some(error) = response.get("error") {
	return Err(anyhow!(error.clone()).context("Error in response"));
}
```
instead of what Vlad asked for earlier:
```
if let Some(error) = response.get("error") {
	anyhow::bail!("Error in response: {}", error)
}
```
in order not to stringify the error, but to preserve the original's Error impl's `source`?

Reply:
> If so, let me point out that the 'error' here is just an arbitrary string field in the json received from UKT.

Good point, I read it as an error, my bad! Ignore that, then.

> Do you want me to turn that into:
>
>     if response.status() != reqwest::StatusCode::OK {
>    	anyhow::bail!("Error in response: {:?}", response.error_for_status())
>     }

Not quite, replacing all of that with:

```
response.error_for_status().context("error in response")?;
```

should work, I believe, since this will automatically bail if it's not `Result::Ok`.

Reply:
Done, although it turned out trickier than expected. TLDR, there were three problems with the suggestion:
- the `anyhow!` macro is not directly compatible with the `Result<Response, Error>` type returned by `response.error_for_status()`,
- the `?` operator can only be applied to values that implement `std::ops::Try`
which is not the case of `anyhow::Error`
- the `response.error_for_status()` consumes the response.

After some digging, I finally came up with this:
```
response.error_for_status_ref().context("received non-OK status code")?;
```